protocols/dcutr/examples/client.rs

```diff
@@ -0,0 +1,275 @@
+        }
+    }
+}
+
+fn main() -> Result<(), Box<dyn Error>> {
+    env_logger::init();
+
+    let opts = Opts::from_args();
+
+    let local_key = generate_ed25519(opts.secret_key_seed);
+    let local_peer_id = PeerId::from(local_key.public());
+    info!("Local peer id: {:?}", local_peer_id);
+
+    let (relay_transport, client) = Client::new_transport_and_behaviour(local_peer_id);
+
+    let noise_keys = noise::Keypair::<noise::X25519Spec>::new()
+        .into_authentic(&local_key)
+        .expect("Signing libp2p-noise static DH keypair failed.");
+
+    let transport = OrTransport::new(
+        relay_transport,
+        block_on(DnsConfig::system(TcpConfig::new().port_reuse(true))).unwrap(),
+    )
+    .upgrade(upgrade::Version::V1)
+    .authenticate(noise::NoiseConfig::xx(noise_keys).into_authenticated())
+    .multiplex(libp2p_yamux::YamuxConfig::default())
+    .boxed();
+
+    #[derive(NetworkBehaviour)]
+    #[behaviour(out_event = "Event", event_process = false)]
+    struct Behaviour {
+        relay_client: Client,
+        ping: Ping,
+        identify: Identify,
+        dcutr: dcutr::behaviour::Behaviour,
+    }
+
+    #[derive(Debug)]
+    enum Event {
+        Ping(PingEvent),
+        Identify(IdentifyEvent),
+        Relay(client::Event),
+        Dcutr(dcutr::behaviour::Event),
+    }
+
+    impl From<PingEvent> for Event {
+        fn from(e: PingEvent) -> Self {
+            Event::Ping(e)
+        }
+    }
+
+    impl From<IdentifyEvent> for Event {
+        fn from(e: IdentifyEvent) -> Self {
+            Event::Identify(e)
+        }
+    }
+
+    impl From<client::Event> for Event {
+        fn from(e: client::Event) -> Self {
+            Event::Relay(e)
+        }
+    }
+
+    impl From<dcutr::behaviour::Event> for Event {
+        fn from(e: dcutr::behaviour::Event) -> Self {
+            Event::Dcutr(e)
+        }
+    }
+
+    let behaviour = Behaviour {
+        relay_client: client,
+        ping: Ping::new(PingConfig::new()),
+        identify: Identify::new(IdentifyConfig::new(
+            "/TODO/0.0.1".to_string(),
+            local_key.public(),
+        )),
+        dcutr: dcutr::behaviour::Behaviour::new(),
+    };
+
+    let mut swarm = SwarmBuilder::new(transport, behaviour, local_peer_id)
+        .dial_concurrency_factor(10_u8.try_into().unwrap())
+        .build();
+
+    swarm
+        .listen_on(
+            Multiaddr::empty()
+                .with("0.0.0.0".parse::<Ipv4Addr>().unwrap().into())
+                .with(Protocol::Tcp(0)),
+        )
+        .unwrap();
+
+    // Wait to listen on localhost.
+    block_on(async {
+        let mut delay = futures_timer::Delay::new(std::time::Duration::from_secs(1)).fuse();
+        loop {
+            futures::select! {
+                            event = swarm.next() => {
+                                match event.unwrap() {
+                                    SwarmEvent::NewListenAddr { address, .. } => {
```

Comment:
Yes, this is misleading. We are giving `Swarm` some time to set up listeners. Though it might be more than one, depending on the number of network interfaces, thus returning after a single `NewListenAddr` might wait for one network interface, but not all.

Waiting for a second instead is a hack, but a hack that works. Happy for alternative suggestions.

At least better documented with 21ae03bbe9a30bc4ce567511f81343d0af54043c.